xla/service/gpu/instruction_fusion_test.cc

```diff
@@ -782,5 +782,54 @@ TEST_F(InstructionFusionTest, IotaIntoVariadicReduction) {
       op::Reduce(op::Parameter(), op::Iota(), op::Constant(), op::Constant()));
 }
 
+TEST_F(InstructionFusionTest, InputReductionFusion) {
+  auto module = ParseAndReturnVerifiedModule(R"(
+    HloModule test_module
+    add.clone.13 {
+      x.27 = f32[] parameter(0)
+      y.27 = f32[] parameter(1)
+      ROOT add.1036 = f32[] add(x.27, y.27)
+    }
+    add.clone.14 {
+      x.28 = f32[] parameter(0)
+      y.28 = f32[] parameter(1)
+      ROOT add.1037 = f32[] add(x.28, y.28)
+    }
+    add {
+      x = bf16[] parameter(0)
+      convert.448 = f32[] convert(x)
+      y = bf16[] parameter(1)
+      convert.449 = f32[] convert(y)
+      add.597 = f32[] add(convert.448, convert.449)
+      ROOT convert.450 = bf16[] convert(add.597)
+    }
+    ENTRY FuseSmallReduction {
+      param_2.7 = bf16[8,16,64,2048]{3,2,1,0} parameter(2)
+      convert.1395 = f32[8,16,64,2048]{3,2,1,0} convert(param_2.7)
+      param_0.85 = bf16[8,16,64,2048]{3,2,1,0} parameter(0)
+      convert.1393 = f32[8,16,64,2048]{3,2,1,0} convert(param_0.85)
+      multiply.1652 = f32[8,16,64,2048]{3,2,1,0} multiply(convert.1395, convert.1393)
+      convert.1392 = bf16[8,16,64,2048]{3,2,1,0} convert(multiply.1652)
+      bitcast.15934 = bf16[128,64,2048]{2,1,0} bitcast(convert.1392)
+      convert.1391 = f32[128,64,2048]{2,1,0} convert(bitcast.15934)
+      param_1.15 = bf16[] parameter(1)
+      convert.1394 = f32[] convert(param_1.15)
+      reduce.462 = f32[128,64]{1,0} reduce(convert.1391, convert.1394), dimensions={2}, to_apply=add.clone.13
+      reduce.121 = f32[64]{0} reduce(reduce.462, convert.1394), dimensions={0}, to_apply=add.clone.14
+      convert.890 = bf16[64]{0} convert(reduce.121)
+      ROOT all-reduce.197 = bf16[64]{0} all-reduce(convert.890), channel_id=213, replica_groups={{0,1,2,3,4,5,6,7}}, use_global_device_ids=true, to_apply=add
+    })")
+                    .value();
+
+  EXPECT_TRUE(duplicating_instruction_fusion_.Run(module.get()).value());
+
+  HloInstruction* root = module->entry_computation()->root_instruction();
+  HloInstruction* fused_convert_fusion = root->mutable_operand(0);
+
+  ASSERT_THAT(fused_convert_fusion, op::Fusion());
+  SCOPED_TRACE(module->ToString());
+  EXPECT_EQ(fused_convert_fusion->fusion_kind(),
+            HloInstruction::FusionKind::kLoop);
```

Comment:
IsReductionFromOrToContiguousDimensions() usually identifies reductions that are roots of input, not loop fusions. Is this here a loop fusion because fusing the convert made it so or because the reduction was elemental even before the fusion?

Reply:
I think i deliberately made this a loop fusion since once we fuse the convert, the reduction is not the root anymore. Do you think this should stay as input fusion?